src/qibolab/designs/abstract.py

```diff
@@ -0,0 +1,50 @@
+from abc import ABC, abstractmethod
+
+
+class AbstractInstrumentDesign(ABC):
+    @abstractmethod
+    def connect(self):
+        """Connect to all instruments."""
+        raise_error(NotImplementedError)
```

Comment:
Not sure why you are using `raise_error` (not a built-in, nor imported), while you could simply:
```suggestion
        raise NotImplementedError
```

However (from [docs](https://docs.python.org/3/library/abc.html#abc.abstractmethod)):
> A class that has a metaclass derived from [ABCMeta](https://docs.python.org/3/library/abc.html#abc.ABCMeta) cannot be instantiated unless all of its abstract methods and properties are overridden.

So, you should not need any statement here, because being abstract not only you can not call it, but you can not even instantiate the class itself, unless the methods are overridden.

True that a method can't be empty, so in principle you might think you need to replace with `pass`. But a docstring is already some content, so you can safely scratch the body completely for as long as you have at least a docstring.